### PR TITLE
3666 auto-filled generated id properties

### LINKED ISSUE #3666: Generated ID field autofill for new properties. (20 pts)
As admins, when we add a generated ID field to a template that already has entities we want that all existing entities get a generated id.

- Test it on large collection
- Do not use the entity save to update all the entities
- Check that the Generated ID is sincronize across languages.

### PR DESCRIPTION
**Pull request by mfacar**
fixes #3666

PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] Test for side effects
- [ ] UI responsiveness
- [ ] Cross browser testing
- [ ] Code review

### CODE REVIEW

**Comment by mfacar** [comment]
After trying different approaches to autofill "generated id" properties the final implementation has the following results in performance.

**Legend:**
- before.- time taken by update without generated id properties 
- remove-prop.- time taken by update after removing a property
- auto-fill.- time taken by update after adding a generated id property

Screen Shot 2021-06-04 at 13 16 46

**Notes**: 
- The number of entities is distributed in three languages
- The time consumed is the result of updates by batches of 1000 entities at a time

There is an important increment of the time over 100000 entities, for that reason I think it could be worth showing a message alerting about the long processing time, what do you think @RafaPolit, @nwyu, @simonfossom

**Comment by Zasa-san** [comment]
Functionality looks good!

File: app/api/entities/generatedIdPropertyAutoFiller.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import { ObjectIdSchema, PropertySchema } from 'shared/types/commonTypes';
+import { generateID } from 'shared/IDGenerator';
+import { propertyTypes } from 'shared/propertyTypes';
+//@ts-ignore
```
[the remark this answers is not in the document]
Comment (mfacar) [reply app/api/entities/generatedIdPropertyAutoFiller.ts:4]: Agree, I tried to update it in this PR but there are some tests that get broken, I opened this Tech debt issue #3741

File: app/api/entities/entities.js [hunk 2]
```diff
@@ -347,6 +347,16 @@ const withDocuments = (entities, documentsFullText, withPdfInfo) =>
     })
   );
 
+const reindexEntitiesByTemplate = async (template, options) => {
+  const templateHasRelationShipProperty = !template.properties?.find(
```
Comment (Zasa-san) [inline app/api/entities/entities.js:351]: Should this be named 'templateHasNorelationshipProperty' ?

File: app/api/entities/entities.js [hunk 3]
```diff
@@ -347,6 +347,16 @@ const withDocuments = (entities, documentsFullText, withPdfInfo) =>
...
+    p => p.type === propertyTypes.relationship
+  );
+  if (options.reindex && (options.generatedIdAdded || templateHasRelationShipProperty)) {
```
Comment (Zasa-san) [inline app/api/entities/entities.js:354]: Assuming that templateHasRelationShipProperty is postponing the reindex. If that's the case then should this be postponed even if generatedIdAdded is true?

File: app/api/templates/specs/generatedIdPropertyAutoFiller.spec.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
+import entities from 'api/entities';
```
Comment (Zasa-san) [inline app/api/templates/specs/generatedIdPropertyAutoFiller.spec.ts:1]: Is there a reason for this test being in a different route than the module is testing?

File: app/api/templates/specs/generatedIdPropertyAutoFiller.spec.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        .map(e => e.metadata!.auto_id![0].value)
+        .filter(unique);
+      expect(differentAutoId.length).toBe(2);
+      const differentAutoId1 = affectedEntities
+        .map(e => e.metadata!.auto_id_1![0].value)
+        .filter(unique);
+      expect(differentAutoId1.length).toBe(2);
+      expect(affectedEntities[0].metadata!.text![0].value).toEqual('test');
+    });
+    it('should reindex updated templates', async () => {
```
Comment (Zasa-san) [inline app/api/templates/specs/generatedIdPropertyAutoFiller.spec.ts:58]: Could this test check that all entities are re-indexed?

File: app/api/templates/specs/templates.spec.js [hunk 6]
```diff
@@ -348,6 +347,37 @@ describe('templates', () => {
...
+        it('should call populateGeneratedIdBTemplate to auto-fill values', async () => {
+          const templateToUpdate = {
+            _id: templateToBeEditedId,
+            name: 'template to be edited',
+            commonProperties: [{ name: 'title', label: 'Title', type: 'text' }],
+            properties: [{ name: 'autoId', type: propertyTypes.generatedid, label: 'Auto Id' }],
+          };
+          await templates.save(templateToUpdate, 'en');
+
+          expect(populateGeneratedIdByTemplateSpy).toHaveBeenCalled();
```
Comment (Zasa-san) [inline app/api/templates/specs/templates.spec.js:370]: Can we change to a toHaveBeenCalledWith() to check that contents are correct?

File: app/api/entities/entities.js (hunk 3 above)
Comment (mfacar) [reply app/api/entities/entities.js:354]: Not really, the postponed reindex don't do a reindex of all the entities, just those entities that were impacted by the changes, for example, the entities without a relationship property, are not reindexed, this is why although it implies a double indexation it's necessary to do it here in order to have all the generated ids indexed.

File: app/api/templates/specs/generatedIdPropertyAutoFiller.spec.ts (hunk 4 above)
Comment (mfacar) [reply app/api/templates/specs/generatedIdPropertyAutoFiller.spec.ts:1]: Yes, it's needed to check the indexation as part of the process. Probably in the future when we address the problem with indexation we can relocate this.

File: app/react/Templates/components/specs/MetadataTemplate.spec.js [hunk 7]
```diff
@@ -228,21 +230,49 @@ describe('MetadataTemplate', () => {
...
+          expect(props.saveTemplate).toHaveBeenCalledWith({
+            ...templateWithId,
+            reindex: true,
+          });
+        });
+
+        describe('when there is a quite amount of entities from the template', () => {
+          it('should ask for a reindex', async () => {
+            await submitTemplate(templateWithId, true, 50000);
+            context.confirm.calls.mostRecent().args[0].cancel();
```
Comment (Zasa-san) [inline app/react/Templates/components/specs/MetadataTemplate.spec.js:266]: Could we add a test case to check when the user click yes?

File: app/api/entities/entities.js (hunk 2 above)
Comment (mfacar) [reply app/api/entities/entities.js:351]: Thanks, I didn't realize that, instead  I extracted the negation, I think it's more readable

File: app/api/templates/specs/generatedIdPropertyAutoFiller.spec.ts (hunk 5 above)
Comment (mfacar) [reply app/api/templates/specs/generatedIdPropertyAutoFiller.spec.ts:58]: Of course, it's done

File: app/api/templates/specs/templates.spec.js (hunk 6 above)
Comment (mfacar) [reply app/api/templates/specs/templates.spec.js:370]: Sure, it's done

File: app/react/Templates/components/specs/MetadataTemplate.spec.js (hunk 7 above)
Comment (mfacar) [reply app/react/Templates/components/specs/MetadataTemplate.spec.js:266]: done

**Pull request merged**